application_sdk/clients/.cursor/BUGBOT.md

```diff
@@ -0,0 +1,280 @@
+- **Client vs Logic separation**: Database clients should handle connections, not business rules
+- **Extract domain logic**: Lock management, caching, or processing logic should be in separate classes
+- **Single purpose interfaces**: Each client class should have one clear responsibility
+
+```python
+# ❌ REJECT: Mixed responsibilities
+class RedisClient:
+    def connect(self):
+        """Handle connection setup"""
+    def acquire_lock(self, lock_name):
+        """Lock functionality - should be separate"""
+    def get_data(self, key):
+        """Client functionality"""
+
+# ✅ REQUIRE: Separated responsibilities
+class RedisClient:
+    def connect(self):
+        """Handle connection setup"""
+    def get_data(self, key):
+        """Client functionality"""
+
+class RedisLockManager:  # Separate class for lock functionality
+    def __init__(self, client: RedisClient):
+        self.client = client
+    def acquire_lock(self, lock_name):
+        """Lock-specific logic"""
+```
+
+**Async Client Patterns:**
+
+- Use async/await for all I/O operations
+- Implement proper connection context managers
+- Handle connection failures gracefully with retries
+- Use asyncio connection pools, not synchronous pools
+
+```python
+# ✅ DO: Proper async connection management
+async def execute_query(self, query: str, params: tuple):
+    async with self.pool.acquire() as conn:
+        try:
+            return await conn.fetch(query, *params)
+        except Exception as e:
+            logger.error(f"Query failed: {query[:100]}...", exc_info=True)
+            raise
+```
+
+**Configuration Management for Clients:**
+
+- **Environment-specific settings**: All connection parameters must be externalized to environment variables
+- **Default value validation**: Every configuration parameter must have a sensible default and validation
+- **Development vs Production**: Client configurations must work in both environments
+- **Configuration consolidation**: Related configuration should be grouped together
+
+```python
+# ✅ DO: Proper client configuration
+class DatabaseClientConfig:
+    def __init__(self):
+        self.host = os.getenv("DB_HOST", "localhost")
+        self.port = int(os.getenv("DB_PORT", "5432"))
+        self.max_connections = int(os.getenv("DB_MAX_CONNECTIONS", "20"))
+        self.timeout = int(os.getenv("DB_TIMEOUT_SECONDS", "30"))
+        self.ssl_required = os.getenv("DB_SSL_REQUIRED", "true").lower() == "true"
+        self._validate()
+
+    def _validate(self):
+        if self.max_connections <= 0:
+            raise ValueError("DB_MAX_CONNECTIONS must be positive")
+        if self.timeout <= 0:
+            raise ValueError("DB_TIMEOUT_SECONDS must be positive")
+
+# ❌ REJECT: Poor configuration management
+class BadDatabaseClient:
+    def __init__(self):
+        self.host = "localhost"  # Hardcoded
+        self.connections = os.getenv("MAX_CONN")  # No default, no validation
+```
```

Comment:
I am not sure where was this picked up from, but in general we centralize all the envs into the `constant.py` file and this is self-contradictory considering earlier in the rule it said otherwise.